### PR TITLE
Feature: Form Builder Onboarding (Tour)

### PR DESCRIPTION
**Pull request by kjohnson**
## Description

This PR adds a form builder onboarding tour using ShepherdJS.

## Visuals

image

Peek 2023-03-22 16-58

Peek 2023-03-28 13-48

## Testing Instructions

Create a new next gen donation form and step through the tour. Once complete, the tour will no longer auto-start on page load. To re-start the tour delete the local storage key `givewp-next-gen-onboarding-complete`.

### CODE REVIEW

**Comment by kjohnson** [comment]
## [Loom] Form Builder Onboarding/Tour

https://www.loom.com/share/f9587b962ef049a281aeeda0f8ecc2a5

image

**Comment by kjohnson** [comment]
[[Loom] [Update] Form Builder Onboarding/Tour](https://www.loom.com/share/6d68fede086e43aba23f6285da3bfe02)

[](https://www.loom.com/share/6d68fede086e43aba23f6285da3bfe02)

**Comment by kjohnson** [comment]
[[Loom] [Update] Form Builder Onboarding/Tour](https://www.loom.com/share/6d68fede086e43aba23f6285da3bfe02)

[](https://www.loom.com/share/6d68fede086e43aba23f6285da3bfe02)

File: packages/form-builder/src/components/onboarding/steps/index.tsx [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            return new Promise<void>(function (resolve) {
+                document.getElementById('AddBlockButtonContainer').querySelector('button').click();
+                resolve();
+            });
+        },
+    },
+    {
+        id: 'editingABlock',
+        attachTo: { element: '.givewp-next-gen-sidebar-primary', on: 'left-start' as Placement },
+        title: 'Editing a block',
```
Comment (Benunc) [inline packages/form-builder/src/components/onboarding/steps/index.tsx:59]:
```suggestion
        title: 'Block Settings',
```

File: packages/form-builder/src/components/onboarding/steps/index.tsx [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        text: 'You can edit the structure of a block in a section. For eg. the “Donation Amount and Levels” is the block being edited here',
```
Comment (Benunc) [inline packages/form-builder/src/components/onboarding/steps/index.tsx:60]:
```suggestion
        text: 'Select a block to edit the settings for that individual block on the Block tab of the editor. Settings will vary depending on the type of block selected.',
```

File: packages/form-builder/src/components/onboarding/steps/index.tsx [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            return new Promise<void>(function (resolve) {
+                document.dispatchEvent(new CustomEvent('selectAmountBlock'));
+                resolve();
+            });
+        },
+    },
+    {
+        id: 'designTab',
+        attachTo: { element: '.components-tab-panel__tabs button:last-of-type', on: 'left' as Placement },
+        title: 'Design tab',
```
Comment (Benunc) [inline packages/form-builder/src/components/onboarding/steps/index.tsx:71]:
```suggestion
        title: 'Design Mode',
```

File: packages/form-builder/src/components/onboarding/steps/index.tsx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        text: 'This is where you get to see and edit the appearance of your form',
```
Comment (Benunc) [inline packages/form-builder/src/components/onboarding/steps/index.tsx:72]:
```suggestion
        text: 'Select the design tab to switch the canvas into a preview of what the form will look like, and access various settings for the visual aspect of the form.',
```

File: packages/form-builder/src/components/onboarding/steps/index.tsx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                // @ts-ignore
+                document.querySelector('.components-tab-panel__tabs button:last-of-type').click();
+                resolve();
+            });
+        },
+    },
+    {
+        id: 'formTemplate',
+        attachTo: { element: '.components-panel__row', on: 'left-start' as Placement },
+        title: 'Form template',
```
Comment (Benunc) [inline packages/form-builder/src/components/onboarding/steps/index.tsx:84]:
```suggestion
        title: 'Form Design',
```

File: packages/form-builder/src/components/onboarding/steps/index.tsx [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        text: 'Select from an array of available form template options here. More template options are being developed now!',
```
Comment (Benunc) [inline packages/form-builder/src/components/onboarding/steps/index.tsx:85]:
```suggestion
        text: 'Select the design of the form based on what you need. More form designs are coming soon!',
```

File: packages/form-builder/src/components/onboarding/steps/index.tsx [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    },
+    {
+        id: 'formDesign',
+        attachTo: { element: 'iframe', on: 'right-start' as Placement },
+        title: 'Form design',
```
Comment (Benunc) [inline packages/form-builder/src/components/onboarding/steps/index.tsx:90]:
```suggestion
        title: 'Live Preview',
```

File: packages/form-builder/src/components/onboarding/steps/index.tsx [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        text: 'This shows how your form looks like with the selected form template',
```
Comment (Benunc) [inline packages/form-builder/src/components/onboarding/steps/index.tsx:91]:
```suggestion
        text: 'As you make changes or select a different design, those changes happen live on the canvas.',
```

File: packages/form-builder/src/components/onboarding/steps/index.tsx [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    },
+    {
+        id: 'editingAFormDesign',
+        attachTo: { element: '.givewp-next-gen-sidebar-primary', on: 'left-start' as Placement },
+        title: 'Editing a form design',
```
Comment (Benunc) [inline packages/form-builder/src/components/onboarding/steps/index.tsx:96]:
```suggestion
        title: 'Design Settings',
```

File: packages/form-builder/src/components/onboarding/steps/index.tsx [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        text: 'You can customize the appearance (i.e. colors and features) of your form based on the selected form template',
```
Comment (Benunc) [inline packages/form-builder/src/components/onboarding/steps/index.tsx:97]:
```suggestion
        text: 'Individual form designs have various settings and options to allow you to customize appearance via  features like goal progress bars, headings, and custom CSS.',
```

File: packages/form-builder/src/components/onboarding/steps/index.tsx [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    },
+    {
+        id: 'congrats',
+        title: 'Congrats 🎉',
+        text: 'Experience the new visual donation form builder and create the donation form of your dreams',
```
Comment (Benunc) [inline packages/form-builder/src/components/onboarding/steps/index.tsx:102]:
```suggestion
        text: 'The visual donation form builder lets you build and customize forms to more easily raise money online. If you need it, access this tour again with the three-dot menu in the top right of the editor screen. Happy fundraising!',
```

File: packages/form-builder/src/components/onboarding/steps/withText.tsx [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    return (
+        <div>
+            {isFirst && (
+                <div style={{display: "flex", justifyContent: 'center', margin: '0 auto var(--givewp-spacing-4)'}}>
+                    <Logo/>
+                </div>
+            )}
+            {!isFirst && !isLast &&(
+                <div style={{display:'flex', backgroundColor: 'var(--givewp-blue-25)', fontSize: '12px',padding:'4px',borderRadius:'2px',justifyContent:'space-between', alignItems: 'center'}}>
+                    <div><strong>Step {stepNumber}</strong> out of {stepCount}</div>
```
Comment (Benunc) [inline packages/form-builder/src/components/onboarding/steps/withText.tsx:16]: this needs to be internationalized. and I think `Step {stepNumber} of {stepCount}` is best.

**Review by Benunc — CHANGES_REQUESTED** [review]
A comment and a bunch of suggestions.

File: packages/form-builder/src/components/onboarding/steps/withText.tsx (hunk 12 above)
Comment (kjohnson) [reply packages/form-builder/src/components/onboarding/steps/withText.tsx]: Resolved.

**Review by Benunc — APPROVED** [review]
Book it.

File: packages/form-builder/src/components/onboarding/Onboarding.tsx [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+            const amountBlockId = amountBlock.getAttribute('data-block');
+            selectBlock(amountBlockId).then(() => console.log('Amount block selected'));
+        }
+
+        document.addEventListener('selectAmountBlock', selectAmountBlockCallback );
+
+        return () => {
+            window.removeEventListener('selectAmountBlock', selectAmountBlockCallback);
+        }
+    })
```
Comment (JasonTheAdams) [inline packages/form-builder/src/components/onboarding/Onboarding.tsx:36]: @kjohnson The way this is written, the effect is going to fire every single time the component renders. Is that the intended behavior?

**Review by JasonTheAdams — CHANGES_REQUESTED** [review]
This is a awesome and I love seeing the design system in use. I had just a couple minor questions.

File: packages/form-builder/src/components/onboarding/Onboarding.tsx (hunk 13 above)
Comment (JasonTheAdams) [reply packages/form-builder/src/components/onboarding/Onboarding.tsx:36]: Am I right in guessing you only want this to fire the first time the component mounts?

Comment (kjohnson) [reply packages/form-builder/src/components/onboarding/Onboarding.tsx:36]: In practice, this doesn't seem to be an issue. This component has no props and doesn't have a reason to re-render.

Comment (kjohnson) [reply packages/form-builder/src/components/onboarding/Onboarding.tsx]: Resolved, see https://github.com/impress-org/givewp-next-gen/pull/143/commits/f54b511974ccf1f77d599ffb746751cf3d6624e7.

**Review by JasonTheAdams — APPROVED** [review]

File: packages/form-builder/src/components/canvas/FormBlocks.tsx [hunk 14]
```diff
@@ -2,13 +2,16 @@ import {BlockList, BlockTools, DefaultBlockAppender, ObserveTyping, WritingFlow}
...
     return (
-        <BlockTools>
-            <WritingFlow>
-                <ObserveTyping>
-                    <BlockList renderAppender={DefaultBlockAppender} />
-                </ObserveTyping>
-            </WritingFlow>
-        </BlockTools>
+        <div id="FormBlocksCanvas" style={{padding: '58px', margin: '-58px'}}>
+            <BlockTools>
```
Comment (jonwaldstein) [inline packages/form-builder/src/components/canvas/FormBlocks.tsx:6]: These styles make me a little uneasy, what's this for?

File: packages/form-builder/src/components/onboarding/steps/index.tsx [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+    }),
+)([
+    {
+        id: 'welcome',
+        title: __('Welcome to the visual donation form builder!', 'give'),
+        text: __('The following is a quick (less than a minute) tour of the visual donation form builder, to introduce the tools for creating engaging donation forms.', 'give'),
+    },
+    {
+        id: 'canvas',
+        attachTo: { element: '#FormBlocksCanvas', on: 'right-start' as Placement },
```
Comment (jonwaldstein) [inline packages/form-builder/src/components/onboarding/steps/index.tsx:28]: I noticed the Id's are being added as camelcase which always feels weird to me as standard convention is typically `form-blocks-canvas` for the ID attribute

File: packages/form-builder/src/components/onboarding/steps/index.tsx [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+    },
+    {
+        id: 'addButton',
+        attachTo: { element: '#AddBlockButtonContainer', on: 'bottom' as Placement },
+        title: __('Sidebar Toggles', 'give'),
+        text: __('These two buttons give you the ability to add and reorder sections and blocks to the canvas, with drag-and-drop ease.', 'give'),
+    },
+    {
+        id: 'addBlock',
+        attachTo: { element: '.givewp-next-gen-sidebar-secondary', on: 'right-start' as Placement },
```
Comment (jonwaldstein) [inline packages/form-builder/src/components/onboarding/steps/index.tsx:47]: One convention I like to use for these types of things is adding classnames with a `js-` prefix so I know that the element is being controlled by js externally and not to change that classname - we might consider using that here.

**Review by jonwaldstein — CHANGES_REQUESTED** [review]

File: packages/form-builder/src/components/canvas/FormBlocks.tsx (hunk 14 above)
Comment (kjohnson) [reply packages/form-builder/src/components/canvas/FormBlocks.tsx:6]: This is specifically to set spacing when highlighting the canvas. There is not actual spacing the space the canvas wrapper from the block list wrapper, so this inline style is used to set the location of the outline.

Comment (kjohnson) [reply packages/form-builder/src/components/canvas/FormBlocks.tsx:6]: Instead, we could target the id and conditionally apply this style when targeted by ShepherdJS.

File: packages/form-builder/src/components/onboarding/steps/index.tsx (hunk 15 above)
Comment (kjohnson) [reply packages/form-builder/src/components/onboarding/steps/index.tsx:28]: I always thought that the convention was camleCase for IDs and kebab-case for classes.

File: packages/form-builder/src/components/onboarding/steps/index.tsx (hunk 16 above)
Comment (kjohnson) [reply packages/form-builder/src/components/onboarding/steps/index.tsx:47]: This is a layered application of JavaScript, though, which is why it is targeting an existing class name.

File: packages/form-builder/src/components/canvas/FormBlocks.tsx (hunk 14 above)
Comment (kjohnson) [reply packages/form-builder/src/components/canvas/FormBlocks.tsx]: I migrated the inline styles to the stylesheet and set the selector to only apply when targeted by ShepherdJS, see https://github.com/impress-org/givewp-next-gen/pull/143/commits/4f5b7de247394ff709d04977590dddb46efb7869.

```css
#FormBlocksCanvas.shepherd-target {
    padding: 58px;
    margin: -58px;
}
```

File: packages/form-builder/src/components/onboarding/steps/withText.tsx [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+            {isFirst && (
+                <div style={{display: "flex", justifyContent: 'center', margin: '0 auto var(--givewp-spacing-4)'}}>
+                    <Logo/>
+                </div>
+            )}
+            {!isFirst && !isLast &&(
+                <div style={{display:'flex', backgroundColor: 'var(--givewp-blue-25)', fontSize: '12px',padding:'4px',borderRadius:'2px',justifyContent:'space-between', alignItems: 'center'}}>
+                    <div>{sprintf(__('Step %s of %s', 'give'), stepNumber, stepCount)}</div>
+                    <Button variant='tertiary' className={'js-exit-tour'}>{__('Exit tour', 'give')}</Button>
+                </div>
```
[the remark this answers is not in the document]
Comment (kjohnson) [reply packages/form-builder/src/components/onboarding/steps/withText.tsx:19]: Hm... perhaps by adjusted step count is off.

[the remark this answers is not in the document]
Comment (kjohnson) [reply packages/form-builder/src/components/onboarding/steps/withText.tsx:19]: I think my step count offset is conflicting with the step index offset.

[the remark this answers is not in the document]
Comment (kjohnson) [reply packages/form-builder/src/components/onboarding/steps/withText.tsx:19]: Resolved, see https://github.com/impress-org/givewp-next-gen/pull/143/commits/ab812ad72628e988ac52599f4ad2582289ba6810.

| First step adjusted for the intro | Last step adjusted for the close |
| -- | --- |
| image | image |

File: packages/form-builder/src/App.scss [hunk 18]
```diff
@@ -480,4 +482,53 @@ This creates a consistent separator between the tabs and the rest of the sidebar
...
+}
+
+.shepherd-element,
+.shepherd-modal-overlay-container {
+    z-index: 9999999999 !important;
+}
+.shepherd-element {
+    padding: var(--givewp-spacing-6);
+    padding-top: var(--givewp-spacing-4);
+}
```
[the remark this answers is not in the document]
Comment (kjohnson) [reply packages/form-builder/src/App.scss]: Resolved, see https://github.com/impress-org/givewp-next-gen/pull/143/commits/c4bc13047c08973623fcda9826041cd927124191.

File: packages/form-builder/src/components/onboarding/steps/index.tsx (hunk 15 above)
Comment (jonwaldstein) [reply packages/form-builder/src/components/onboarding/steps/index.tsx:28]: Technically both is valid, I personally prefer the latter & if camelCase the first letter should be lowercase - but looking at how WP they are using kebab

Screenshot 2023-03-31 at 1 27 04 PM

Comment (kjohnson) [reply packages/form-builder/src/components/onboarding/steps/index.tsx]: Resolved, see https://github.com/impress-org/givewp-next-gen/pull/143/commits/3b33f9cbb341fba69412e53d6943eee70d8ec77d.

File: packages/form-builder/src/components/onboarding/steps/index.tsx (hunk 16 above)
Comment (jonwaldstein) [reply packages/form-builder/src/components/onboarding/steps/index.tsx:47]: yeah i'm really just thinking this classname & others will probably end up changing (for example: to remove the word `next-gen`) & if we update a gutenberg package and the underlying classname we're targeting changes then this stops working.  Just something to think about for longevity

File: packages/form-builder/src/components/canvas/FormBlocks.tsx (hunk 14 above)
Comment (jonwaldstein) [reply packages/form-builder/src/components/canvas/FormBlocks.tsx]: okay cool, what happens if you don't have these styles?

Comment (kjohnson) [reply packages/form-builder/src/components/canvas/FormBlocks.tsx]: The boarder collapses, which makes it difficult to distinguish between steps one and two.

File: packages/form-builder/src/components/onboarding/steps/index.tsx (hunk 16 above)
Comment (kjohnson) [reply packages/form-builder/src/components/onboarding/steps/index.tsx]: Sure. I switched these to target IDs instead.

**Review by jonwaldstein — APPROVED** [review]
👍

**Pull request merged**